src/providers/mixed-route-quote-provider.ts

```diff
@@ -0,0 +1,845 @@
+import { BigNumber } from '@ethersproject/bignumber';
```

Comment:
This file has little to no changes from `v3QuoteProovider`.

Here's a diff between the two (sdiff -s):
```
eric.zhong@UNISWAP-MAC-020 src % sdiff -s providers/mixed-route-quote-provider.ts providers/v3/quote-provider.ts 
import { encodeMixedRouteToPath } from '@uniswap/router-sdk'; |	import { encodeRouteToPath } from '@uniswap/v3-sdk';
import retry from 'async-retry';			      |	import retry, { Options as RetryOptions } from 'async-retry';
import { MixedRoute } from '../routers/router';		      |	import { V3Route } from '../../routers/router';
import { IQuoterV3__factory } from '../types/other/factories/ |	import { IQuoterV2__factory } from '../../types/v3/factories/
import { ChainId, metric, MetricLoggerUnit, QUOTER_V2_ADDRESS |	import { ChainId, metric, MetricLoggerUnit } from '../../util
import { CurrencyAmount } from '../util/amounts';	      |	import { QUOTER_V2_ADDRESS } from '../../util/addresses';
import { log } from '../util/log';			      |	import { CurrencyAmount } from '../../util/amounts';
import { routeToString } from '../util/routes';		      |	import { log } from '../../util/log';
import { Result } from './multicall-provider';		      |	import { routeToString } from '../../util/routes';
import { UniswapMulticallProvider } from './multicall-uniswap |	import { Result } from '../multicall-provider';
import { ProviderConfig } from './provider';		      |	import { UniswapMulticallProvider } from '../multicall-uniswa
							      >	import { ProviderConfig } from '../provider';
 * @dev import shared types and classes from v3QuoteProvider  |	 * A quote for a swap on V3.
import {						      |	export type V3AmountQuote = {
  BatchParams,						      <
  BlockConflictError,					      <
  BlockNumberConfig,					      <
  FailureOverrides,					      <
  ProviderBlockHeaderError,				      <
  ProviderGasError,					      <
  ProviderTimeoutError,					      <
  QuoteRetryOptions,					      <
  SuccessRateError,					      <
} from './v3/quote-provider';				      <
							      <
/**							      <
 * A on chain quote for a mixed route swap (both V2 and V3 po <
 */							      <
export type MixedRouteAmountQuote = {			      <
							      >	export class BlockConflictError extends Error {
							      >	  public name = 'BlockConflictError';
							      >	}
							      >	export class SuccessRateError extends Error {
							      >	  public name = 'SuccessRateError';
							      >	}
							      >
							      >	export class ProviderBlockHeaderError extends Error {
							      >	  public name = 'ProviderBlockHeaderError';
							      >	}
							      >
							      >	export class ProviderTimeoutError extends Error {
							      >	  public name = 'ProviderTimeoutError';
							      >	}
							      >
							      >	/**
							      >	 * This error typically means that the gas used by the multic
							      >	 * exceeded the total call gas limit set by the node provider
							      >	 *
							      >	 * This can be resolved by modifying BatchParams to request f
							      >	 * quotes per call, or to set a lower gas limit per quote.
							      >	 *
							      >	 * @export
							      >	 * @class ProviderGasError
							      >	 */
							      >	export class ProviderGasError extends Error {
							      >	  public name = 'ProviderGasError';
							      >	}
							      >
							      >	export type QuoteRetryOptions = RetryOptions;
							      >
 * The mixed route and a list of quotes for that route.	      |	 * The V3 route and a list of quotes for that route.
export type MixedRouteWithQuotes = [MixedRoute, MixedRouteAmo |	export type V3RouteWithQuotes = [V3Route, V3AmountQuote[]];
 * Provider for getting mixed route quotes on Uniswap.	      |	 * Provider for getting quotes on Uniswap V3.
 * @interface IMixedRouteQuoteProvider			      |	 * @interface IV3QuoteProvider
export interface IMixedRouteQuoteProvider {		      |	export interface IV3QuoteProvider {
   * For every route, gets an exactIn quote for every amount  |	   * For every route, gets an exactIn quotes on V3 for every 
   * @returns For each route returns a MixedRouteWithQuotes o |	   * @returns For each route returns a V3RouteWithQuotes obje
    routes: MixedRoute[],				      |	    routes: V3Route[],
  ): Promise<{						      |	  ): Promise<{ routesWithQuotes: V3RouteWithQuotes[]; blockNu
    routesWithQuotes: MixedRouteWithQuotes[];		      |
    blockNumber: BigNumber;				      |	  /**
  }>;							      |	   * For every route, gets ane exactOut quote on V3 for every
							      >	   *
							      >	   * @param amountOuts The amounts to get quotes for.
							      >	   * @param routes The routes to get quotes for.
							      >	   * @param [providerConfig] The provider config.
							      >	   * @returns For each route returns a V3RouteWithQuotes obje
							      >	   * @returns The blockNumber used when generating the quotes
							      >	   */
							      >	  getQuotesManyExactOut(
							      >	    amountOuts: CurrencyAmount[],
							      >	    routes: V3Route[],
							      >	    providerConfig?: ProviderConfig
							      >	  ): Promise<{ routesWithQuotes: V3RouteWithQuotes[]; blockNu
							      >	/**
							      >	 * The parameters for the multicalls we make.
							      >	 *
							      >	 * It is important to ensure that (gasLimitPerCall * multical
							      >	 *
							      >	 * V3 quotes can consume a lot of gas (if the swap is so larg
							      >	 * number of ticks), so there is a risk of exceeded gas limit
							      >	 */
							      >	export type BatchParams = {
							      >	  /**
							      >	   * The number of quotes to fetch in each multicall.
							      >	   */
							      >	  multicallChunk: number;
							      >	  /**
							      >	   * The maximum call to consume for each quote in the multic
							      >	   */
							      >	  gasLimitPerCall: number;
							      >	  /**
							      >	   * The minimum success rate for all quotes across all multi
							      >	   * If we set our gasLimitPerCall too low it could result in
							      >	   * quotes failing due to out of gas. This parameters will f
							      >	   * in this case.
							      >	   */
							      >	  quoteMinSuccessRate: number;
							      >	};
							      >
							      >	/**
							      >	 * The fallback values for gasLimit and multicallChunk if any
							      >	 *
							      >	 */
							      >
							      >	export type FailureOverrides = {
							      >	  multicallChunk: number;
							      >	  gasLimitOverride: number;
							      >	};
							      >
							      >	export type BlockHeaderFailureOverridesDisabled = { enabled: 
							      >	export type BlockHeaderFailureOverridesEnabled = {
							      >	  enabled: true;
							      >	  // Offset to apply in the case of a block header failure. e
							      >	  rollbackBlockOffset: number;
							      >	  // Number of batch failures due to block header before tryi
							      >	  attemptsBeforeRollback: number;
							      >	};
							      >	export type BlockHeaderFailureOverrides =
							      >	  | BlockHeaderFailureOverridesDisabled
							      >	  | BlockHeaderFailureOverridesEnabled;
							      >
							      >	/**
							      >	 * Config around what block number to query and how to handle
							      >	 */
							      >	export type BlockNumberConfig = {
							      >	  // Applies an offset to the block number specified when fet
							      >	  // Useful for networks where the latest block may not be av
							      >	  baseBlockOffset: number;
							      >	  // Config for handling header not found errors.
							      >	  rollback: BlockHeaderFailureOverrides;
							      >	};
							      >
 * Computes quotes for mixed routes. For mixed routes, quotes |	 * Computes quotes for V3. For V3, quotes are computed on-cha
 * the 'QuoterV3' smart contract. This is because computing q |	 * the 'QuoterV2' smart contract. This is because computing q
 * exceed this limit. Additionally, each quote on chain can c |	 * exceed this limit. Additionally, each quote on V3 can cons
 * @class MixedRouteQuoteProvider			      |	 * @class V3QuoteProvider
export class MixedRouteQuoteProvider implements IMixedRouteQu |	export class V3QuoteProvider implements IV3QuoteProvider {
   * Creates an instance of MixedRouteQuoteProvider.	      |	   * Creates an instance of V3QuoteProvider.
      : undefined; /// for now				      |	      : QUOTER_V2_ADDRESS;
        `No address for Uniswap QuoterV3 Contract on chain id |	        `No address for Uniswap QuoterV2 Contract on chain id
    if (quoterAddress === QUOTER_V2_ADDRESS)		      <
      throw new Error('QuoterV2 is not supported for mixed pa <
							      <
    routes: MixedRoute[],				      |	    routes: V3Route[],
    routesWithQuotes: MixedRouteWithQuotes[];		      |	    routesWithQuotes: V3RouteWithQuotes[];
							      >	  public async getQuotesManyExactOut(
							      >	    amountOuts: CurrencyAmount[],
							      >	    routes: V3Route[],
							      >	    providerConfig?: ProviderConfig
							      >	  ): Promise<{
							      >	    routesWithQuotes: V3RouteWithQuotes[];
							      >	    blockNumber: BigNumber;
							      >	  }> {
							      >	    return this.getQuotesManyData(
							      >	      amountOuts,
							      >	      routes,
							      >	      'quoteExactOutput',
							      >	      providerConfig
							      >	    );
							      >	  }
							      >
    routes: MixedRoute[],				      |	    routes: V3Route[],
    functionName: 'quoteExactInput',			      |	    functionName: 'quoteExactInput' | 'quoteExactOutput',
    routesWithQuotes: MixedRouteWithQuotes[];		      |	    routesWithQuotes: V3RouteWithQuotes[];
        const encodedRoute = encodeMixedRouteToPath(	      |	        const encodedRoute = encodeRouteToPath(
          false // we don't support exactOut for mixedRoute p |	          functionName == 'quoteExactOutput' // For exactOut 
                    contractInterface: IQuoterV3__factory.cre |	                    contractInterface: IQuoterV2__factory.cre
      .flatMap((routeWithQuotes: MixedRouteWithQuotes) => rou |	      .flatMap((routeWithQuotes: V3RouteWithQuotes) => routeW
    routes: MixedRoute[],				      |	    routes: V3Route[],
  ): MixedRouteWithQuotes[] {				      |	  ): V3RouteWithQuotes[] {
    const routesQuotes: MixedRouteWithQuotes[] = [];	      |	    const routesQuotes: V3RouteWithQuotes[] = [];
      const quotes: MixedRouteAmountQuote[] = _.map(	      |	      const quotes: V3AmountQuote[] = _.map(
        `Failed quotes for mixed routes Part ${idx}/${Math.ce |	        `Failed quotes for V3 routes Part ${idx}/${Math.ceil(
eric.zhong@UNISWAP-MAC-020 src % 
```